source/volume_of_fluid/solver.cc

```diff
@@ -0,0 +1,118 @@
+{
+  template <int dim>
+  void VolumeOfFluidHandler<dim>::solve_volume_of_fluid_system (const VolumeOfFluidField<dim> field)
+  {
+    unsigned int block_idx = field.volume_fraction.block_index;
+
+    sim.computing_timer.enter_section ("   Solve volume of fluid system");
+    this->get_pcout() << "   Solving volume of fluid system... " << std::flush;
+
+    const double tolerance = std::max(1e-50,
+                                      volume_of_fluid_solver_tolerance*sim.system_rhs.block(block_idx).l2_norm());
+
+    SolverControl solver_control (1000, tolerance);
+
+#ifdef ASPECT_USE_PETSC
+    SolverCG<LinearAlgebra::Vector> solver(solver_control);
+    LinearAlgebra::PreconditionJacobi precondition;
+    precondition.initialize(sim.system_matrix.block(block_idx, block_idx));
+#else
+    TrilinosWrappers::SolverCG solver(solver_control);
+    TrilinosWrappers::PreconditionJacobi precondition;
+    precondition.initialize(sim.system_matrix.block(block_idx, block_idx));
+#endif
+
+    // Create distributed vector (we need all blocks here even though we only
+    // solve for the current block) because only have a ConstraintMatrix
+    // for the whole system, current_linearization_point contains our initial guess.
+    LinearAlgebra::BlockVector distributed_solution (
+      this->introspection().index_sets.system_partitioning,
+      this->get_mpi_communicator());
+    distributed_solution.block(block_idx) = sim.current_linearization_point.block (block_idx);
+
+    sim.current_constraints.set_zero(distributed_solution);
+
+    // solve the linear system:
+    try
+      {
+        solver.solve (sim.system_matrix.block(block_idx,block_idx),
+                      distributed_solution.block(block_idx),
+                      sim.system_rhs.block(block_idx),
+                      precondition);
+      }
+    // if the solver fails, report the error from processor 0 with some additional
+    // information about its location, and throw a quiet exception on all other
+    // processors
+    catch (const std::exception &exc)
+      {
+        if (Utilities::MPI::this_mpi_process(this->get_mpi_communicator()) == 0)
+          AssertThrow (false,
+                       ExcMessage (std::string("The iterative advection solver "
+                                               "did not converge. It reported the following error:\n\n")
+                                   +
+                                   exc.what()))
+          else
+            throw QuietException();
+      }
+
+    sim.current_constraints.distribute (distributed_solution);
+    sim.solution.block(block_idx) = distributed_solution.block(block_idx);
+
+    // print number of iterations and also record it in the
+    // statistics file
+    this->get_pcout() << solver_control.last_step()
+                      << " iterations." << std::endl;
+
+    // Do not add VolumeOfFluid solver iterations to statistics, duplicaiton due to
+    // splitting messes with file format
```

Comment:
Please explain a bit further `messes`, does it lead to duplicated or empty lines in the statistics file?
Also `duplicaiton`-> duplication

Reply:
If I remember correctly, it produced "duplicated" lines where many of the statistics were missing (and possibly was inconsistent about which line had the actual data).

Reply:
Then what about:
```
    // Do not add VolumeOfFluid solver iterations to statistics, because the repeated
    // execution for different directions within one time step duplicates lines in the output file.
```